src/dune_engine/target_promotion.ml

```diff
@@ -103,91 +103,86 @@ let delete_stale_dot_merlin_file ~dir ~source_files_to_ignore =
   in
   source_files_to_ignore
 
-let promote ~dir ~targets ~promote ~promote_source =
-  let open Fiber.O in
-  let { Rule.Promote.lifetime; only; into; _ } = promote in
-  (* CR-someday amokhov: Don't ignore directory targets. *)
-  let file_targets =
-    Targets.map targets ~f:(fun ~files ~dirs ->
-        ignore dirs;
-        files)
+let promote ~dir ~targets_and_digests ~promote ~promote_source =
+  let selected_for_promotion =
+    match promote.Rule.Promote.only with
+    | None -> fun (_ : Path.Build.t) -> true
+    | Some pred ->
+      fun target ->
+        Predicate_lang.Glob.exec pred ~standard:Predicate_lang.any
+          (Path.reach ~from:(Path.build dir) (Path.build target))
+  in
+  let relocate =
+    match promote.into with
+    | None -> Fun.id
+    | Some { loc; dir } ->
+      fun target_in_source_tree ->
+        Path.Source.relative
+          (Path.Source.relative
+             (Path.Source.parent_exn target_in_source_tree)
+             dir ~error_loc:loc)
+          (Path.Source.basename target_in_source_tree)
   in
-  Fiber.parallel_iter_set
-    (module Path.Build.Set)
-    file_targets
-    ~f:(fun target ->
-      let consider_for_promotion =
-        match only with
-        | None -> true
-        | Some pred ->
-          Predicate_lang.Glob.exec pred
-            (Path.reach (Path.build target) ~from:(Path.build dir))
-            ~standard:Predicate_lang.any
+  let open Fiber.O in
+  (* CR-someday amokhov: When promoting directory targets, we might want to
+     create the destination directory instead of reporting an error. Maybe we
+     should just always do that? After all, if the user says "promote results
+     into this directory, please", they might expect Dune to create it too. *)
```

Comment:
@rgrinberg I added this comment and then found your PR #3073 that added the logic to error out if the `into` directory doesn't exist. I wonder what you think about relaxing this conservative approach and instead creating the `into` directory if it doesn't exist.

Reply:
The reason I ask is that when promoting directory targets, we probably want to create all (sub)directories because otherwise it would probably be quite annoying for the user: they would have to `mkdir` a set of directories for the promotion to work, and that set may keep changing as the rule changes.

Reply:
Oh, I guess a problem here is that creating directories in the source tree would trigger rebuilds.

Reply:
> Oh, I guess a problem here is that creating directories in the source tree would trigger rebuilds.

Is that so bad? I thought we were going to remove all the special handling around polling and promotion.

Reply:
Yeah, with the eager cutoff layer, it's probably OK.